### PR TITLE
Reduce gradual audio desyncing in dumps and apply the correct sample rate for GameCube audio

### PR DESCRIPTION
**Pull request by CasualPokePlayer**
This PR aims to resolve issue 12951. Its current status right now should reduce the gradual desync, but not completely eliminate it. Eliminating it would be a bit complicated here, as either audio dumps would need to be reworked into some format that accepts non-integer sample rates (which I haven't has any luck finding such), or resampling the audio to an integer sample rate (probably the most "simple" here). Keeping this as draft for now.

### CODE REVIEW

**Comment by AdmiralCurtiss** [comment]
Your 54000000 * 2 divisor is quite a ways towards the int32 limit, you might need to update a few of these variables to be int64.

**Comment by CasualPokePlayer** [comment]
> Your 54000000 * 2 dividend is quite a ways towards the int32 limit, you might need to update a few of these variables to be int64.

The dividend is a u64 in the first place, integer promotion should handle issues of that limit.

**Comment by JosJuice** [comment]
@dolphin-emu-bot rebuild

**Comment by CasualPokePlayer** [comment]
Not sure what to do now. Seems like resampling is the only solution here, nothing seems to support a non-integer sample rate, and the resampling code used in the Mixer is a bit hard to follow.

**Comment by Pokechu22** [comment]
@dolphin-emu-bot rebuild

**Comment by JosJuice** [comment]
I think I would prefer to just output without resampling and set a rounded sample rate in the WAV header, at least for now. I don't think everyone wants the resampling, and it is possible to fix the incorrect sample rate that's specified in the WAV header in post.

**Comment by CasualPokePlayer** [comment]
Alright, reverte those hacky resampling commits and just have the other fixes. This should at least make the GC sample rate "accurate" I guess. Resampling can come at some later point in time.

File: Source/Core/Core/HW/AudioInterface.h [hunk 1]
```diff
@@ -22,12 +22,12 @@ bool IsPlaying();
 
 void RegisterMMIO(MMIO::Mapping* mmio, u32 base);
 
-// Get the audio rates (48000 or 32000 only)
-u32 GetAIDSampleRate();
-u32 GetAISSampleRate();
+// Get the audio rate divisors (divisors for 48KHz or 32KHz only)
+u32 GetAIDSampleRateDivisor();
+u32 GetAISSampleRateDivisor();
```
Comment (JosJuice) [inline Source/Core/Core/HW/AudioInterface.h:27]: Could we put some kind of comment here saying that `Mixer::FIXED_SAMPLE_RATE_DIVIDEND` is the dividend in use? This feels a bit "magic" otherwise.

We could also create a fraction struct and pass around both the dividend and divisor everywhere, but maybe that would be overkill...

File: Source/Core/Core/HW/DVD/DVDInterface.cpp [hunk 2]
```diff
@@ -294,12 +294,14 @@ static u32 AdvanceDTK(u32 maximum_samples, u32* samples_to_process)
 static void DTKStreamingCallback(DIInterruptType interrupt_type, const std::vector<u8>& audio_data,
                                  s64 cycles_late)
 {
-  // TODO: Should we use GetAISSampleRate instead of a fixed 48 KHz? The audio mixer is using
-  // GetAISSampleRate. (This doesn't affect any actual games, since they all set it to 48 KHz.)
-  const u32 sample_rate = AudioInterface::Get48KHzSampleRate();
+  // Actual games always set this for 48 KHz
+  // although if it were changed to 32KHz, it would likely wreak havoc on audio code
```
Comment (JosJuice) [inline Source/Core/Core/HW/DVD/DVDInterface.cpp:298]: This line of this comment is a little confusing. I'm assuming that what you mean is that if we use `AudioInterface::Get48KHzSampleRate`, there would be havoc if the game changed to 32 KHz, but once this is merged and this comment is read without the context of the changeset, the connection to the now removed call to `AudioInterface::Get48KHzSampleRate` would be unclear. Maybe change the comment to something like "Actual games always set this to 48 KHz, but let's make sure to use GetAISSampleRateDivisor() just in case it changes to 32 KHz". (Feel free to reword)

File: Source/Core/AudioCommon/Mixer.h [hunk 3]
```diff
@@ -51,6 +51,10 @@ class Mixer final
   float GetCurrentSpeed() const { return m_speed.load(); }
   void UpdateSpeed(float val) { m_speed.store(val); }
 
+  static constexpr u64 FIXED_SAMPLE_RATE_DIVIDEND =
+      54000000 *
+      2;  // 54000000 doesn't work here as it doesn't evenly divide with 32000, but 108000000 does
```
Comment (JosJuice) [inline Source/Core/AudioCommon/Mixer.h:56]: Please put the comment on the line above the variable definition. It will make the formatting neater.

File: Source/Core/Core/HW/DVD/DVDInterface.cpp [hunk 4]
```diff
@@ -294,12 +294,14 @@ static u32 AdvanceDTK(u32 maximum_samples, u32* samples_to_process)
...
+  // so let's make sure to use GetAISSampleRateDivisor()
+  const u32 sample_rate_divisor = AudioInterface::GetAISSampleRateDivisor();
 
   // Determine which audio data to read next.
-  const u32 maximum_samples = sample_rate / 2000 * 7;  // 3.5 ms of samples
+  const u32 maximum_samples =
+      Mixer::FIXED_SAMPLE_RATE_DIVIDEND / 2000 * 7 / sample_rate_divisor;  // 3.5 ms of samples
```
Comment (JosJuice) [inline Source/Core/Core/HW/DVD/DVDInterface.cpp:304]: Same here regarding putting the comment on a separate line.

File: Source/Core/Core/HW/SystemTimers.cpp [hunk 5]
```diff
@@ -110,7 +111,8 @@ void DSPCallback(u64 userdata, s64 cyclesLate)
 int GetAudioDMACallbackPeriod()
 {
   // System internal sample rate is fixed at 32KHz * 4 (16bit Stereo) / 32 bytes DMA
-  return s_cpu_core_clock / (AudioInterface::GetAIDSampleRate() * 4 / 32);
+  return static_cast<u64>(s_cpu_core_clock) * AudioInterface::GetAIDSampleRateDivisor() /
+         (Mixer::FIXED_SAMPLE_RATE_DIVIDEND * 4 / 32);
```
Comment (JosJuice) [inline Source/Core/Core/HW/SystemTimers.cpp:115]: Minor performance optimization: You should be able to avoid the division by 32 here by instead multiplying with 32 on the left-hand side of the equation.

File: Source/Core/Core/HW/AudioInterface.cpp [hunk 6]
```diff
@@ -122,8 +122,8 @@ void DoState(PointerWrap& p)
   p.Do(s_sample_counter);
   p.Do(s_interrupt_timing);
   p.Do(s_last_cpu_time);
-  p.Do(s_ais_sample_rate);
-  p.Do(s_aid_sample_rate);
+  p.Do(s_ais_sample_rate_divisor);
+  p.Do(s_aid_sample_rate_divisor);
```
Comment (JosJuice) [inline Source/Core/Core/HW/AudioInterface.cpp:126]: Please update the savestate version in State.cpp, otherwise loading an old savestate will completely mess up the sample rates.

File: Source/Core/AudioCommon/WaveFile.cpp [hunk 7]
```diff
@@ -26,7 +27,7 @@ WaveFileWriter::~WaveFileWriter()
   Stop();
 }
 
-bool WaveFileWriter::Start(const std::string& filename, unsigned int HLESampleRate)
+bool WaveFileWriter::Start(const std::string& filename, unsigned int HLESampleRateDivisor)
```
Comment (JosJuice) [inline Source/Core/AudioCommon/WaveFile.cpp:30]: Since you're renaming this variable anyway, could you change it so it matches our style guide? (`hle_sample_rate_divisor`)

**Comment by JosJuice** [comment]
@dolphin-emu-bot rebuild

File: Source/Core/Core/HW/SystemTimers.cpp (hunk 5 above)
Comment (CasualPokePlayer) [reply Source/Core/Core/HW/SystemTimers.cpp:115]: Mixer::FIXED_SAMPLE_RATE_DIVIDEND is a constant expression, so shouldn't the compiler optimize that / 32 out?

Comment (JosJuice) [reply Source/Core/Core/HW/SystemTimers.cpp:115]: That's true. You can skip my suggestion then.

**Comment by JosJuice** [comment]
@dolphin-emu-bot rebuild

**Comment by Pokechu22** [comment]
@dolphin-emu-bot rebuild

**Comment by CasualPokePlayer** [comment]
Looking through audio formats again and realizing there are formats that support non-integer sample rates, one namely being the aiff format. Perhaps this could be used as a solution here.

**Comment by Filoppi** [comment]
@CasualPokePlayer https://github.com/dolphin-emu/dolphin/pull/8983 this PR introduced floating point sampling rate support for the audio backend. Dumps were still approximated to the closest integer as it didn't really mattered IMO. I'm still trying to get back and finish it given the only pending thing was finishing the improved audio stretching code. It might be outdated and require some merging, but I think most of the changes would still apply. This is the first time in two years I have some time off work so I might finally go back and finish it.

**Comment by AdmiralCurtiss** [comment]
@dolphin-emu-bot rebuild

**Comment by CasualPokePlayer** [comment]
The old "appoximates" just do not work here, as it causes issues with the actual audio timing code (rounding issues causes it to be off from its reported rate), which causes noticable audio desyncs within at least an hour. To quote from the relevant issue
```
https://github.com/dolphin-emu/dolphin/blob/d32b13fb36e6616f7bb4a5a395c34ae2f2ada32b/Source/Core/Core/HW/SystemTimers.cpp#L113
This is the relevant timing code. It expects a constant number of cycles per sample, doing math with the CPU clock rate and such to figure that out. The divisions do not divide evenly, causing rounding errors and the actual output sample rate to be around 32024.2hz/48061.7hz.
```

The changes in this PR still have audio resampled for the backend (as it was previously doing) and that resampler supported input floats anyways, but at least its original input sample rate will be accurate (and not have errors like reported 32029 -> actual 32024.2).

**Comment by AdmiralCurtiss** [comment]
Yeah, this PR is kinda orthogonal to any actual audio backend changes. This actually affects the timing.

**Comment by JosJuice** [comment]
Could we do this AIFF stuff in a separate PR afterwards, for ease of reviewing?

**Comment by CasualPokePlayer** [comment]
Fair enough

**Comment by JosJuice** [comment]
@dolphin-emu-bot rebuild

**Review by AdmiralCurtiss — COMMENTED** [review]
Some minor stuff but this looks correct to me.

**Comment by AdmiralCurtiss** [comment]
@dolphin-emu-bot rebuild

**Review by AdmiralCurtiss — APPROVED** [review]
